src/rbac/default_role_manager.rs

```diff
@@ -0,0 +1,353 @@
+}
+
+impl DefaultRoleManager {
+    pub fn new(max_hierarchy_level: usize) -> Self {
+        return DefaultRoleManager {
+            all_roles: HashMap::new(),
+            max_hierarchy_level,
+            has_pattern: false,
+            matching_func: None,
+        };
+    }
+
+    fn create_role(&mut self, name: &str) -> Rc<RefCell<Role>> {
+        self.all_roles
+            .entry(name.to_owned())
+            .or_insert(Rc::new(RefCell::new(Role::new(name.to_owned()))))
+            .clone()
+    }
+
+    fn has_role(&self, name: &str) -> bool {
+        if let Some(_role) = self.all_roles.get(name) {
+            return true;
+        }
+        return false;
+    }
+}
+
+impl RoleManager for DefaultRoleManager {
+    fn clone_box(&self) -> Box<dyn RoleManager> {
+        Box::new(self.clone())
+    }
+
+    fn add_link(&mut self, name1: &str, name2: &str, domain: Vec<&str>) {
+        let mut name1 = name1.to_owned();
+        let mut name2 = name2.to_owned();
+        if domain.len() == 1 {
+            name1 = format!("{}::{}", domain[0], name1);
+            name2 = format!("{}::{}", domain[0], name2);
+        } else if domain.len() > 1 {
+            panic!("error domain length");
+        }
+        let role1 = self.create_role(&name1);
+        let role2 = self.create_role(&name2);
+        role1.borrow_mut().add_role(Rc::clone(&role2));
+    }
+
+    fn delete_link(&mut self, name1: &str, name2: &str, domain: Vec<&str>) {
+        let mut name1 = name1.to_owned();
+        let mut name2 = name2.to_owned();
+        if domain.len() == 1 {
+            name1 = format!("{}::{}", domain[0], name1);
+            name2 = format!("{}::{}", domain[0], name2);
+        } else if domain.len() > 1 {
+            panic!("error domain length");
+        }
+        if !self.has_role(&name1) || !self.has_role(&name2) {
+            panic!("name12 error");
+        }
+        let role1 = self.create_role(&name1);
+        let role2 = self.create_role(&name2);
+        role1.borrow_mut().delete_role(Rc::clone(&role2));
+    }
+
+    fn has_link(&mut self, name1: &str, name2: &str, domain: Option<&str>) -> bool {
+        let mut name1 = name1.to_owned();
+        let mut name2 = name2.to_owned();
+        if let Some(domain_val) = domain {
+            name1 = format!("{}::{}", domain_val, name1);
+            name2 = format!("{}::{}", domain_val, name2);
+        }
+        if name1 == name2 {
+            return true;
+        }
+        if !self.has_role(&name1) || !self.has_role(&name2) {
+            return false;
+        }
+        let role1 = self.create_role(&name1);
+        return role1.borrow().has_role(&name2, self.max_hierarchy_level);
+    }
+
+    fn get_roles(&mut self, name: &str, domain: Option<&str>) -> Vec<String> {
+        let mut name = name.to_owned();
+        if let Some(domain_val) = domain {
+            name = format!("{}::{}", domain_val, name);
+        }
+        if !self.has_role(&name) {
```

Comment:
name is already a reference. This call takes a double reference. The compiler will coerce double references to normal references so this code compiles, but it is confusing to read.

Reply:
Nope, `let mut name = name.to_owned();` will convert a `&str` into `String`, so `&name` is not a double reference but a `&str`. I'm writting this to concat name with domain, maybe I should use `concat!` instead.